starboard/shared/libiamf/iamf_audio_decoder.cc

```diff
@@ -0,0 +1,382 @@
+}
+
+bool IamfAudioDecoder::DecodeInternal(
+    const scoped_refptr<InputBuffer>& input_buffer) {
+  SB_DCHECK(BelongsToCurrentThread());
+  SB_DCHECK(input_buffer);
+  SB_DCHECK(input_buffer->size() > 0);
+  SB_DCHECK(output_cb_);
+  SB_DCHECK(!stream_ended_ || !pending_audio_buffers_.empty());
+  SB_DCHECK(is_valid());
+
+  reader_.ResetAndRead(input_buffer);
+  if (!reader_.is_valid()) {
+    SB_DLOG(INFO) << "Failed to parse IA Descriptors";
+    error_cb_(kSbPlayerErrorDecode, "Failed to parse IA Descriptors");
+    return false;
+  }
+  if (!decoder_is_configured_) {
+    if (!InitializeCodec()) {
+      SB_DLOG(INFO) << "Failed to initialize IAMF decoder";
+      error_cb_(kSbPlayerErrorDecode, "Failed to initialize IAMF decoder");
+      return false;
+    }
+  }
+
+  scoped_refptr<DecodedAudio> decoded_audio = new DecodedAudio(
+      audio_stream_info_.number_of_channels, GetSampleType(),
+      kSbMediaAudioFrameStorageTypeInterleaved, input_buffer->timestamp(),
+      audio_stream_info_.number_of_channels * reader_.samples_per_buffer() *
+          starboard::media::GetBytesPerSample(GetSampleType()));
+  int samples_decoded = IAMF_decoder_decode(
+      decoder_, reader_.data().data(), reader_.data().size(), nullptr,
+      reinterpret_cast<void*>(decoded_audio->data()));
+  if (samples_decoded < 1) {
+    SB_DLOG(INFO) << "IAMF_decoder_decode() error "
+                  << ErrorCodeToString(samples_decoded);
+    error_cb_(kSbPlayerErrorDecode, "Failed to decode IAMF sample, error " +
+                                        ErrorCodeToString(samples_decoded));
+    return false;
+  }
+
+  SB_DCHECK(samples_decoded <= reader_.samples_per_buffer());
+
+  decoded_audio->ShrinkTo(audio_stream_info_.number_of_channels *
+                          reader_.samples_per_buffer() *
+                          starboard::media::GetBytesPerSample(GetSampleType()));
+
+  // TODO: Enable partial audio once float32 pcm output is available.
+  const auto& sample_info = input_buffer->audio_sample_info();
+  decoded_audio->AdjustForDiscardedDurations(
+      audio_stream_info_.samples_per_second,
+      sample_info.discarded_duration_from_front,
+      sample_info.discarded_duration_from_back);
+
+  decoded_audios_.push(decoded_audio);
+
+  output_cb_();
+
+  return true;
+}
+
+void IamfAudioDecoder::WriteEndOfStream() {
+  SB_DCHECK(BelongsToCurrentThread());
+  SB_DCHECK(output_cb_);
+
+  stream_ended_ = true;
+  if (!pending_audio_buffers_.empty()) {
+    return;
+  }
+
+  // Put EOS into the queue.
+  decoded_audios_.push(new DecodedAudio);
+
+  Schedule(output_cb_);
+}
+
+bool IamfAudioDecoder::InitializeCodec() {
+  SB_DCHECK(is_valid());
+  SB_DCHECK(!decoder_is_configured_);
+
+  // TODO: libiamf has an issue outputting 32 bit float samples, set to 16 bit
+  // for now.
+  int error = IAMF_decoder_set_bit_depth(decoder_, 16);
+  if (error != IAMF_OK) {
+    SB_DLOG(ERROR) << "IAMF_decoder_set_bit_depth() fails with error "
```

Comment:
Let's consider to report the error via "error_cb_" with the details instead of a common error message in InitializeCodec().